falcon/_typing.py

```diff
@@ -104,6 +104,9 @@ async def __call__(
 HeaderMapping = Mapping[str, str]
 HeaderIter = Iterable[Tuple[str, str]]
 HeaderArg = Union[HeaderMapping, HeaderIter]
+
+NarrowHeaderArg = Union[Dict[str,str], List[Tuple[str, str]]]
```

Comment:
I don't understand why we need this. it's more specific that `HeaderArg`

Reply:
Ok, the reason is this code: 

```python
def _load_headers(headers: Optional[HeaderArg]) -> Headers:
    """Transform the headers to dict."""
    if headers is None:
        return {}
    if isinstance(headers, dict):
        return headers
    return dict(headers)
```

The header args were wide-typed as either Mapping[str, str] or Iterable[tuple[str, str]]. The isinstance(..., dict) is not enough to tell Mapping from Iterable[tuple]. I.e. `for k,v in {('foo', 'bar') : 42, ('a', 'b'): 111}` yields perfect `('foo', 'bar'), ('a', 'b')` pairs matching the Iterable[tuple[str, str]]. That's the pyright complain. 

The docs state the arguments to errors must be list or dict (not list-like or dict-like), I thought it would be the right thing to match the typing to runtime.

Reply:
this seems overly pedantic. it's likely fine to narrow `HeaderArg` to use `Sequence[Tuple...]` that should solve the complain

Reply:
I changed the the NarrowHeaderArg to Union[Mapping[], Sequence[]], pyright is ok with it.
Changing the very HeaderArg from Iterable to Sequence is wrong imho, since it's used in `set_headers`.
The docs state the set_headers accepts iterable of tuples but sequence will disallow using generators there.